Security observability

nwmac opened the pull request.

nwmac committed:
Work in progress Analysis tools

nwmac committed:
Thursday updates

nwmac committed:
Friday fixes and improvements

nwmac committed:
Friday code

nwmac committed:
Merge remote-tracking branch 'origin/master' into hackweek-analysis-tools

nwmac committed:
Separate analyzers into a separate container

nwmac committed:
Merge remote-tracking branch 'origin/master' into hackweek-analysis-tools

nwmac committed:
Wire in analyzers container in the Helm chart

nwmac committed:
Hide analysis UI features when not enabled

nwmac committed:
Fix sidepanel bug with fallback metadata

nwmac committed:
Bug fix for change in way tab links are hidden

nwmac committed:
Remove debug logging

nwmac committed:
Add refresh button to report selector drop down. Change no reports icon

nwmac committed:
Add support for adding breadcrumbs in the sub nav bar

nwmac committed:
Fix unit tests

nwmac committed:
Fix format issues

nwmac committed:
Final front-end unit test fixes

nwmac committed:
Fix issues when deploying via Helm with mariadb

nwmac committed:
Analyzers container fix. Allow helm chart to be packaged.

nwmac committed:
Build script fixes

nwmac committed:
Remove file

nwmac committed:
Merge remote-tracking branch 'origin/master' into hackweek-analysis-tools

nwmac committed:
WIP: Add support for Clair image scanning

nwmac committed:
Use klar

nwmac committed:
Remove binary

nwmac committed:
Add clair helm chart for dev

nwmac committed:
Merge branch 'scanners-clair' of github.com:SUSE/stratos into scanners-clair

nwmac committed:
Fixes

nwmac committed:
USe end var for clair server address

nwmac committed:
Latest updates

nwmac committed:
Improvements

nwmac committed:
Minor fixes

nwmac committed:
Tweak

nwmac committed:
Fix 1.16 detecton issue with the analyzers

nwmac committed:
Chart fixes

nwmac committed:
Merge branch 'scanners-clair' of github.com:SUSE/stratos into scanners-clair

richard-cox committed:
Merge remote-tracking branch 'upstream/v4-a9-build' into v4-build

richard-cox committed:
Changes following first run of script

richard-cox committed:
Changes following npm install

richard-cox committed:
Update custom-src to new model
- expose custom module's module's
- Add routing module
- Tweak stratos.config.ts log output
- remove custom-src dir

richard-cox committed:
update naming... custom extensions --> suse extensions

nwmac committed:
Merge remote-tracking branch 'origin/merge-upstream' into scanners-clair

nwmac committed:
Merge remote-tracking branch 'origin/master' into scanners-clair

nwmac committed:
A few tidyups to help review

nwmac committed:
Fix build issue due to merge

richard-cox committed:
Merge remote-tracking branch 'upstream/master' into v4-build

richard-cox committed:
Fixes following merge from upstream

richard-cox committed:
Merge remote-tracking branch 'upstream/master' into v4-build

nwmac committed:
Remove clair from this PR

richard-cox committed:
Ignore example packages when there's a stratos config file

richard-cox committed:
Merge remote-tracking branch 'upstream/master' into merge-upstream

richard-cox committed:
Changes following review

richard-cox committed:
Merge remote-tracking branch 'upstream/master' into merge-upstream

richard-cox committed:
Merge remote-tracking branch 'upstream/theming-improvements' into merge-upstream

richard-cox committed:
Merge remote-tracking branch 'upstream/ignore-example-packages-when-config' into merge-upstream

richard-cox committed:
Merge remote-tracking branch 'origin/merge-upstream' into v4-build

# Conflicts:
#	README.md
#	custom-src/deploy/kubernetes/customize-helm.sh
#	package-lock.json
#	src/frontend/packages/store/src/base-entity-types.ts
#	src/jetstream/go.sum
#	src/tsconfig.json

richard-cox committed:
Changes following merge

richard-cox committed:
Update dir names, remove examples folder

richard-cox committed:
Add back in custom-src deploy content, also add product version to config

richard-cox committed:
Revert change needed downstream... (only needed when suse extension is included)

nwmac committed:
Remove unused wip report viewers

nwmac committed:
Merge remote-tracking branch 'origin/master' into security-observability

richard-cox committed:
Merge remote-tracking branch 'origin/master' into merge-upstream

richard-cox committed:
Fix after merge

richard-cox committed:
Merge remote-tracking branch 'origin/merge-upstream' into v4-build

nwmac committed:
Merge remote-tracking branch 'origin/merge-upstream'

nwmac committed:
Merge remote-tracking branch 'origin/v4-build'

nwmac committed:
Merge remote-tracking branch 'origin/security-observability'

nwmac committed:
Merge branch 'neil' into security-observability

richard-cox committed:
Move new terminal & config code to plugin, fix more build files

nwmac committed:
Fix imports and add doc

nwmac committed:
Merge remote-tracking branch 'origin/v4-build' into security-observability

nwmac committed:
Fix compilation issues

richard-cox committed:
Merge remote-tracking branch 'upstream/master' into merge-upstream

richard-cox committed:
Change following merge

richard-cox committed:
Merge remote-tracking branch 'origin/merge-upstream' into v4-build

nwmac committed:
Tweaks to logging

nwmac committed:
Fix bug where report can not be deleted

richard-cox committed:
Fix kube config connect after merge, also fix subtype & error on connect

richard-cox committed:
Fix e2e

nwmac committed:
Improve drop-down menu

richard-cox committed:
Remove strange merge artifacts

nwmac committed:
Merge remote-tracking branch 'origin/v4-build' into security-observability

nwmac committed:
Remove build file

nwmac committed:
Merge remote-tracking branch 'origin/master' into security-observability

nwmac committed:
Fix graph overview

richard-cox commented on src/frontend/packages/core/src/shared/components/cards/card-number-metric/card-number-metric.component.theme.scss:
```diff
@@ -9,4 +9,16 @@
+
+  .number-metric-card__alert-badge {
+    &-error {
+      background-color: red;
```
We've got some status colours that would be good to use here
```
  $status-colors: map-get($app-theme, status);
  $status-success: map-get($status-colors, success);
  $status-warning: map-get($status-colors, warning);
  $status-danger: map-get($status-colors, danger);
  $status-tentative: map-get($status-colors, tentative);
  $status-info: map-get($status-colors, info);
```

richard-cox commented on src/frontend/packages/suse-extensions/src/custom/kubernetes/kubernetes-entity-factory.ts:
```diff
@@ -104,6 +105,13 @@ entityCache[kubernetesDashboardEntityType] = new KubernetesEntitySchema(
+entityCache[analysisReportEntityType] = new KubernetesEntitySchema(
+  analysisReportEntityType,
+  {},
+  { idAttribute: (entity) => entity.id }
```
This can be just `{ idAttribute: 'id' }`

richard-cox commented on src/frontend/packages/suse-extensions/src/custom/kubernetes/kubernetes-resource-viewer/kubernetes-resource-viewer.component.ts:
```diff
@@ -44,32 +45,41 @@ export class KubernetesResourceViewerComponent implements PreviewableComponent {
+
   setProps(props: KubernetesResourceViewerConfig) {
     this.title = props.title;
+    this.analysis = props.analysis;
```
This is probably a v2 thing, but would be cool to tie this into the pods lists as well.
- They should all use a common base config service (`BaseKubernetesPodsListConfigService`)
- Would need an extra column to show some kind of summary state
- Would need `analysis` property passed through in name cell's cellConfig 
- The overlay enable/selection drop down integration might be tricky. Maybe the kube analysis service could store this and be provided/have an instance per place it's used?

richard-cox commented on src/frontend/packages/suse-extensions/src/custom/kubernetes/list-types/analysis-reports-list-config.service.ts:
```diff
@@ -0,0 +1,129 @@
+
+  private listActionDelete: IListAction<any> = {
+    action: (item) => {
+      console.log(item);
```
console.log (there's a few dotted around)

richard-cox commented on src/frontend/packages/suse-extensions/src/custom/kubernetes/list-types/analysis-reports-list-source.ts:
```diff
@@ -0,0 +1,50 @@
+    endpointService: KubernetesEndpointService,
+    ngZone: NgZone,
+  ) {
+    const action = new GetAnalysisReports(endpointService.baseKube.guid);
```
This i guess should be action builder'ed. Also the endpoint guid is being ignored

richard-cox commented on src/frontend/packages/suse-extensions/src/custom/kubernetes/list-types/analysis-reports-list-source.ts:
```diff
@@ -0,0 +1,50 @@
+      store,
+      action,
+      schema: kubernetesEntityFactory(analysisReportEntityType),
+      getRowUniqueId: (entity: AnalysisReport) => entity.id,
```
For consistency this can be grabbed from the schema in the action

richard-cox commented on src/frontend/packages/suse-extensions/src/custom/kubernetes/services/kubernetes.analysis.service.ts:
```diff
@@ -0,0 +1,305 @@
+    const del = this.http.delete(url, requestArgs);
+    del.subscribe(d => {
+      const action = new GetAnalysisReports(this.kubeEndpointService.baseKube.guid);
+      this.store.dispatch(new ClearPaginationOfType(action));
```
I _think_ the `ResetPagination` action should work better here, think that's the one that retains user filter/sort choices

richard-cox commented on src/frontend/packages/suse-extensions/src/custom/kubernetes/services/kubernetes.analysis.service.ts:
```diff
@@ -0,0 +1,305 @@
+
+  public refresh() {
+    const action = new GetAnalysisReports(this.kubeEndpointService.baseKube.guid);
+    this.store.dispatch(new ClearPaginationOfType(action));
```
Same action could be used here

richard-cox commented on src/frontend/packages/suse-extensions/src/custom/kubernetes/services/kubernetes.analysis.service.ts:
```diff
@@ -0,0 +1,305 @@
+      } else {
+        msg = `${type} analysis started for the Kubernetes cluster`;
+      }
+      const ref = this.snackBar.open(msg, 'View', { duration: 5000 });
```
This should use the snackbar service

richard-cox commented on src/frontend/packages/suse-extensions/src/custom/kubernetes/services/kubernetes.analysis.service.ts:
```diff
@@ -0,0 +1,305 @@
+        kubeScoreHelper.map();
+        break;
+      default:
+        console.log('Do not know how to handle this report type');
```
This should be console.warning or similar, same with all error handling in this service

richard-cox commented on src/frontend/packages/suse-extensions/src/custom/kubernetes/store/kubernetes.effects.ts:
```diff
@@ -97,6 +99,38 @@ export class KubernetesEffects {
+      const entityKey = entityCatalog.getEntityKey(action);
+      return this.http
+        .get(url, requestArgs)
+        .pipe(mergeMap(response => {
```
This needs a `catchError` which returns a `WrapperRequestActionFailed`

richard-cox commented on src/frontend/packages/suse-extensions/src/custom/kubernetes/tabs/kubernetes-analysis-tab/kubernetes-analysis-info/analysis-info-card/analysis-info-card.component.ts:
```diff
@@ -0,0 +1,55 @@
+  }
+
+  constructor(private http: HttpClient) {
+    this.renderer.link = (href, title, text) => `<a target="_blank" title="${title}" href="${href}">${text}</a>`;
```
This title is `null`

richard-cox commented on src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/release/tabs/helm-release-resource-graph/helm-release-resource-graph.component.html:
```diff
@@ -3,23 +3,31 @@
-    <span>Toggle Layout</span>
-  </button>
+
+  <app-analysis-report-selector *ngIf="analyzerService.enabled$ | async" (selected)="analysisChanged($event)" [endpoint]="helper.endpointGuid" [path]="path"></app-analysis-report-selector>
```
This doesn't show the analysis reports for me. Selecting 'none' throws 
```
core.js:6228 ERROR TypeError: ctx_r3.analysisChanged is not a function
    at HelmReleaseResourceGraphComponent_app_analysis_report_selector_6_Template_app_analysis_report_selector_selected_0_listener (helm-release-resource-graph.component.html:7)
```

richard-cox commented on src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/release/tabs/helm-release-resource-graph/helm-release-resource-graph.component.ts:
```diff
@@ -79,6 +83,11 @@ export class HelmReleaseResourceGraphComponent implements OnInit, OnDestroy {
       this.links = newLinks;
       this.update$.next(true);
+
+      // if (!this.didInitialFit) {
```
Comments

richard-cox reviewed (changes requested):
I've jumped the gun a bit and gone through the frontend code (backend it's probably better for you to talk through). Added comments to https://github.com/SUSE/stratos/issues/401 also.

nwmac committed:
Merge remote-tracking branch 'origin/icons' into security-observability

nwmac committed:
Numerous improvements to graph parsing and presentation

nwmac committed:
Remove logging. Add no reports message to workload analysis

nwmac committed:
Add support for CRDs. KubeCF renders correctly.

nwmac committed:
Allow which engines are enabled to be configured

richard-cox committed:
Minor changes following review

nwmac committed:
Fix issue where reports are not filtered by endpoint

nwmac committed:
Fixes for a few more issues

nwmac committed:
Add Analyzers image build to Concourse CI

richard-cox committed:
Multiple small fixes
- fix text search in analysis list
- fix title of links in analysers info page
- handle slow connections by only polling analysis list when not already

richard-cox committed:
Fix kubeGuid for helm world

richard-cox committed:
Add AnalysisReportRunnerComponent
- Still need to add this to other places

nwmac committed:
Delete reports when endpoint is unregistered

nwmac committed:
Merge branch 'security-observability' of github.com:SUSE/stratos into security-observability

nwmac committed:
Buf fixes. Use breadcrumbs in sub-nav

richard-cox committed:
Add run analysis button to workload analyis and graph tabs

richard-cox committed:
Fix select of overlay in workload graphs page

richard-cox committed:
Change default sort order of analysis list to age

richard-cox committed:
Ensure table cell links update on row change

richard-cox committed:
Align table action's icon better

nwmac committed:
Use a side panel for analyzer info

richard-cox committed:
Add actions/effects for all used analyis actions
- Add new ResetPaginationOfType action, like ResetPagination but applies to all types
- Allows user to refresh reports list after kicking off new report on namespace & workload tabs
- Handle missing report param in reports returned from get all reports

richard-cox committed:
Remove some console.logs, converted some to console.info

richard-cox committed:
Update Kube Dashboard, allow download link to be configurable
- Default download link updated to v2.0.3
- Can configured link by setting env var `STRATOS_KUBERNETES_DASHBOARD_IMAGE`
- Can configure env var in helm via `console.kubeDashboardImage`
- Kube Dashboard now expanded by default (to show namespace drop down)

richard-cox committed:
Merge remote-tracking branch 'origin/master' into update-kube-dashboard-url

richard-cox committed:
Fix after merge

richard-cox committed:
Merge remote-tracking branch 'origin/master' into update-kube-dashboard-url

richard-cox committed:
Changes following review

richard-cox committed:
Fix expand of kube dashboard header by default

richard-cox committed:
Merge remote-tracking branch 'origin/master' into security-observability

richard-cox committed:
Changes following review

richard-cox committed:
Merge remote-tracking branch 'origin/update-kube-dashboard-url' into security-observability

# Conflicts:
#	package-lock.json
#	src/frontend/packages/suse-extensions/package.json
#	src/jetstream/default.config.properties

nwmac commented on deploy/kubernetes/console/values.yaml:
```diff
@@ -117,6 +121,7 @@ images:
   fdbserver: stratos-fdbserver
   fdbdoclayer: stratos-fdbdoclayer
   chartsync: stratos-chartsync
+  analyzers: stratos-analyzers
```
Remove

nwmac replied:
or use a default

nwmac commented on src/frontend/packages/core/src/features/dashboard/dashboard-base/dashboard-base.component.ts:
```diff
@@ -133,8 +135,9 @@ export class DashboardBaseComponent implements OnInit, OnDestroy, AfterViewInit
-    );
+      this.tabNavService.tabSubNav$,
+      this.tabNavService.tabSubNavBreadcrumbs$
+    ).pipe(map(([tabNav, tabSubNav, tabSubNavBreadcrumb]) => [tabNav ? tabNav.label : null, tabSubNav, tabNav, tabSubNavBreadcrumb]));
```
Get rid of data[2] and always return [tabNav, tabSubNav, tabSubNavBreadcrumb]

nwmac commented on src/frontend/packages/suse-extensions/src/custom/kubernetes/analysis-report-viewer/analysis-report-selector/analysis-report-selector.component.html:
```diff
@@ -0,0 +1,17 @@
+    <button (click)="refreshReports($event)" mat-menu-item>Refresh</button>
+  </mat-menu>
+
+  <!-- Note - it would be good to only show if there's no reports `*ngIf="analyzers.length > 0"`. However this causes
```
Create an issue for this @richard-cox

nwmac commented on src/frontend/packages/suse-extensions/src/custom/kubernetes/analysis-report-viewer/analysis-report-viewer.component.ts:
```diff
@@ -0,0 +1,75 @@
+  private id: string;
+
+  @Input('report')
+  set report(report: any) {
```
Add typing

richard-cox committed:
Fix json-viewer dark mode

richard-cox committed:
Fix profile page and side nav top position following header diet
- Fix side nav top position
- Update fix for profile page to also work in non-desktop mode

richard-cox committed:
WIP Wire in alerts to workload graph
- need to understand if namespace should be checked when matching node/resource to alert
- need to apply correct colour

richard-cox committed:
Merge remote-tracking branch 'origin/master' into security-observability

richard-cox committed:
Fix workload security analysis overlay slide in

richard-cox committed:
Merge remote-tracking branch 'origin/master' into security-observability

richard-cox committed:
Hide analysis headers info in tech preview & tie in tech preview check to analysisService.hideAnalysis$
- Q should the backend plugins be available in tech preview, see TODO

richard-cox committed:
Hide the Workload Graph view if in tech preview

richard-cox commented on src/jetstream/plugins/analysis/container/kubescore.go:
```diff
@@ -0,0 +1,62 @@
+
+	scriptPath := filepath.Join(getScriptFolder(), "kubescore-runner.sh")
+	args := []string{scriptPath, job.KuebConfigPath, job.Config.Namespace}
+	log.Error(scriptPath)
```
Could be debug

richard-cox commented on src/jetstream/plugins/analysis/container/kubescore.go:
```diff
@@ -0,0 +1,62 @@
+	setJobNameAndPath(job, "Kube-score")
+
+	scriptPath := filepath.Join(getScriptFolder(), "kubescore-runner.sh")
+	args := []string{scriptPath, job.KuebConfigPath, job.Config.Namespace}
```
`KuebConfigPath` typo

richard-cox commented on src/jetstream/plugins/analysis/container/kubescore.go:
```diff
@@ -0,0 +1,62 @@
+			log.Error(err)
+			job.Status = "error"
+		} else {
+			reportFile := filepath.Join(job.Folder, "repor.log")
```
`repor.log`

richard-cox replied:
CI - When out of tech preview - clean old copies

richard-cox commented on src/jetstream/plugins/analysis/container/popeye.go:
```diff
@@ -0,0 +1,115 @@
+	job.Format = "popeye"
+	setJobNameAndPath(job, "Popeye")
+
+	log.Warn("Path is %s", job.Path)
```
log.Warn & others

richard-cox commented on src/jetstream/plugins/analysis/container/routes.go:
```diff
@@ -0,0 +1,88 @@
+	id := ec.Param("id")
+	folder := filepath.Join(a.reportsDir, user, endpoint, id)
+	if err := os.RemoveAll(folder); err != nil {
+		log.Warnf("Could not delete Analysis report folder: %s", folder)
```
Return err or log text to user. Same below

richard-cox commented on src/jetstream/plugins/analysis/container/run.go:
```diff
@@ -0,0 +1,129 @@
+	}
+
+	if err != nil {
+		log.Error("Error running analyzer: %s", err)
```
Add `job.Status = "error"`

richard-cox commented on src/jetstream/plugins/analysis/container/status.go:
```diff
@@ -0,0 +1,71 @@
+		}
+
+		for _, id := range cleanup {
+			log.Errorf("Removing job >>>> %s", id)
```
log.Errorf

richard-cox commented on src/jetstream/plugins/analysis/main.go:
```diff
@@ -0,0 +1,134 @@
+
+// Init performs plugin initialization
+func (analysis *Analysis) Init() error {
+	log.Info("Analysis plugin loaded")
```
Gate on tech preview (check other plugins). Check with line 94

richard-cox commented on src/jetstream/plugins/analysis/run.go:
```diff
@@ -0,0 +1,197 @@
+	defer rsp.Body.Close()
+	response, err := ioutil.ReadAll(rsp.Body)
+	if err != nil {
+		report.Status = "error"
```
report.Status here and line 170 not needed

richard-cox commented on src/jetstream/plugins/analysis/run.go:
```diff
@@ -0,0 +1,197 @@
+	// Job submitted okay
+	// Updated job is in the response
+
+	defer rsp.Body.Close()
```
Confirm when this should be called (only if ReadAll?)

richard-cox commented on src/jetstream/plugins/analysis/run.go:
```diff
@@ -0,0 +1,197 @@
+	return ec.JSON(200, report)
+}
+
+func getScriptFolder() string {
```
Is this still needed?

richard-cox commented on src/jetstream/plugins/analysis/status.go:
```diff
@@ -0,0 +1,102 @@
+
+	ids := make([]string, 0)
+	for _, job := range running {
+		log.Infof("Got running job: %s", job.ID)
```
should be debug

richard-cox commented on src/jetstream/plugins/analysis/status.go:
```diff
@@ -0,0 +1,102 @@
+			if err := dbStore.UpdateReport(job.UserID, job); err != nil {
+				log.Warnf("Unable to update status for job %s: %v", job.ID, err)
+			}
+		}
```
Review clean up process. Add else case, mark job as error (in db, but not in container)?

richard-cox committed:
Fix disable of analysis plugin when tech preview is switched off

richard-cox reviewed (commented):
see comments

nwmac committed:
Adderss PR feedback

richard-cox commented on src/jetstream/plugins/kubernetes/helm/release.go:
```diff
@@ -74,18 +78,33 @@ func (r *HelmRelease) parseManifest() {
 	reader := bytes.NewReader([]byte(r.Manifest))
 	buffer := bufio.NewReader(reader)
 	var bufr strings.Builder
+	fe, _ := os.Create("/home/nwm/manifest_err.yaml")
```
Debug code

richard-cox committed:
Minor tidy ups, fix analysis in graph
- apply typing to many places
- handle kube resources that we fail to fetch/parse
- wire in analysis overlay to graphs and resource slide in

richard-cox committed:
Remove debug code

richard-cox committed:
Merge remote-tracking branch 'origin/master' into security-observability

richard-cox committed:
Merge remote-tracking branch 'origin/master' into security-observability
